Cleanup

gehel opened the pull request:
* Minor cleanup of the ChatMonitorCommandExecutor
* More extensive refactoring of WordManager

Some problems remain on those 2 classes (see individual commit messages).

I'm fairly confident that I did not break anything, but I'm not familiar enough with the whole codebase to be completely sure. Please review carefully before merging!

Commit:
Minor cleanup of ChatMonitorCommandExecutor.

* Extract only the portion of ChatMonitor that needs to be exposed
  (reducing visibility)
* Reducing the scope of some local variables
* Move validation logic to use exception

Note that more work is needed to manage the exceptional flow via
exception, but it makes sense to do a cleanup of WordManager first
and come back to this later.

Commit:
Attempt to simplify WordManager.

* Extracted the loading of word list to a separate class
* Limited the number of different ways of storing words, since they
  can be derived from each others
* Moved to immutable data structure (one step closer to thread-safety)

This refactoring shows that there are still issues with thread-safety
and with the management of `null`. A number of methods, parameters and
fields have been tagged with @Nullable and @Nonnull to clarify the
current situation.

The constructor has some problematic exception handling, which might
lead to a half-initialized instance. A better understanding of the
plugin lifecycle is needed to propose a coherent solution.

mooeypoo commented on src/main/java/io/github/mooeypoo/chatmonitor/words/WordManager.java:
```diff
@@ -1,71 +1,64 @@
+			// Refresh all configs
 			this.configManager.reload();
+			// Redo word collection
+			setWords(wordLoader.collectWords());
```
Random style question -- should this (and others in the class, around) not be `this.setWords` since it's a reference to a class method, rather than a parameter inside this method? I know both work, I'm just curious about the "best practices" of when to use `this.` and when not to, as it seems to be different than what I'd usually be used to...?

gehel replied:
I think being explicit about this made sense 15 years ago when IDE were crap. Nowadays, your IDE will color instance variables and local variables differently, so it should be obvious which is which. And it should provide good enough navigation that even if you need to see the implementation, it should be trivial to get there. In most cases, you shouldn't care anyway where a method is defined when reading the code, it's intent should be clear enough.

In the end, it's a question of personal preference. I don't care much either way. I tend to not use `this` unless it is required  (an instance variable being hidden by a local variable, like in a constructor).

Commit:
Simplification of ConfigManager.

* Removed a few unused methods to ensure that internal representation
  does not leak in the future.
* Fully load configuration at reload time to ensure that accessing
  the configuration afterward does not run into IO or
  ConfigurationExceptions.
* Expose word groups as a Set of Map.Entry to ensure coherent access
  to the groups and limit the number of lookups required.
* Minor other random fixes.

Commit:
Minor cleanups as reported by Spotbugs.

* Explicit use of UTF-8 where needed.
* Proper closing of resources.
* Ensuring that some collections are unmodifiable.
